isis/src/base/apps/framestitch/framestitch.cpp

```diff
@@ -0,0 +1,220 @@
+      throw IException(IException::User, msg, _FILEINFO_);
+    }
+
+    bool inputFlipped = false;
+
+    if (evenCube->hasGroup("Instrument") && oddCube->hasGroup("Instrument")) {
+      const PvlGroup &evenInst = evenCube->group("Instrument");
+      const PvlGroup &oddInst = oddCube->group("Instrument");
+      // Use the start time as an indicator of being the same original image
+      if (QString::compare(QString(evenInst["StartTime"]), QString(oddInst["StartTime"])) != 0) {
+        QString msg = "Even and odd cube start times must match.";
+        throw IException(IException::User, msg, _FILEINFO_);
+      }
+
+      if (evenInst.hasKeyword("DataFlipped") && oddInst.hasKeyword("DataFlipped")) {
+        if (toBool(evenInst["DataFlipped"]) != toBool(oddInst["DataFlipped"])) {
+          QString msg = "Both input cubes must be flipped or not flipped. Cannot combine "
+                        "a flipped and unflipped cube.";
+          throw IException(IException::User, msg, _FILEINFO_);
+        }
+        inputFlipped = toBool(evenInst["DataFlipped"]);
+      }
+    }
+
+    QString outCubeFile = ui.GetCubeName("TO");
+    Cube *outCube = process.SetOutputCube(
+          outCubeFile, CubeAttributeOutput(outCubeFile),
+          evenCube->sampleCount(), evenCube->lineCount(), evenCube->bandCount());
+
+    int frameSize = 1;
+    if (ui.WasEntered("FRAMESIZE")) {
+      frameSize = ui.GetInteger("FRAMESIZE");
+    }
+    else {
+      // User didn't pass the size of the frames so attempt to infer it from
+      // the cubes
+      int evenFrameSize = computeFrameSize(evenCube);
+      int oddFrameSize = computeFrameSize(oddCube);
+
+      if (evenFrameSize == 0) {
+        QString msg = "Failed to find any NULL frames in even cube when attempting "
+                      "to compute frame size.";
+        throw IException(IException::User, msg, _FILEINFO_);
+      }
+      if (oddFrameSize == 0) {
+        QString msg = "Failed to find any NULL frames in odd cube when attempting "
+                      "to compute frame size.";
+        throw IException(IException::User, msg, _FILEINFO_);
+      }
+      if (evenFrameSize != oddFrameSize) {
+        QString msg = "Computed frame sizes for even cube [" + toString(evenFrameSize)
+                    + "] and odd cube [" + toString(oddFrameSize) + "] do not match.";
+        throw IException(IException::User, msg, _FILEINFO_);
+      }
+
+      frameSize = evenFrameSize;
+    }
+
+    // If there's an even number of frames and the inputs are flipped, we have to
+    // swap even and odd because the first frame in the even cube is valid and the
+    // first frame in the odd cube is now all NULL.
+    //
+    //  Before   --Flip-->  After
+    // Even  Odd          Even  Odd
+    // 0000  ####         ####  0000
+    // ####  0000         0000  ####
+    // 0000  ####         ####  0000
+    // ####  0000         0000  ####
+    //
+    int numFrames = evenCube->lineCount() / frameSize;
```

Comment:
I think this implicitly assumes that all frames have the same number of lines (frameSize). I am also a little confused here. I would think that lineCount() includes both the lines which have nulls (which look black) and the lines which have valid data, no?

Reply:
Yes this is true. All frames will have the same number of lines. We want to include all of the lines, including the NULL frames, because we will need to iterate through every frame.

Reply:
I see. So then the frame size includes both valid and invalid lines. I think that was not clear from the comment on top of that function. I assumed only valid lines are counted.